packages/account/src/test-utils/launchNode.ts

```diff
@@ -242,13 +213,26 @@ export const launchNode = async ({
       });
     }
 
-    const cleanupConfig: KillNodeParams = {
-      child,
-      configPath: tempDir,
-      killFn: treeKill,
-      state: {
-        isDead: false,
-      },
+    const childState = {
+      isDead: false,
+    };
+
+    const cleanup = () => {
+      if (childState.isDead) {
+        return;
+      }
+      childState.isDead = true;
+
+      // Remove all the listeners we've added.
+      child.stderr.removeAllListeners();
+
+      // Remove the temporary folder and all its contents.
+      if (existsSync(tempDir)) {
+        rmSync(tempDir, { recursive: true });
+      }
+
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      process.kill(-child.pid!);
```

Comment:
We probably should log an error if we are unable to kill the process

```suggestion
if (child.pid !== undefined) {
  process.kill(-child.pid);
} else {
  console.error('No PID available for the child process, unable to kill launchTestNode');
}
```

Reply:
From the [docs](https://nodejs.org/api/child_process.html#subprocesspid) of `child.pid`:
> Returns the process identifier (PID) of the child process. If the child process fails to spawn due to errors, then the value is undefined and error is emitted.

This will never be `undefined` because `cleanup` will only be returned and be callable if the `child` has started successfully and the node started properly ([logic](https://github.com/FuelLabs/fuels-ts/blob/aeb7ff8bc9d6fc04cbadd063cdb7ab84b8682900/packages/account/src/test-utils/launchNode.ts#L251)). That's why I think there's no need for this type of proposed handling.

Reply:
I agree with your suggestion at face value but I also feel that there's nothing to lose in having that over disabling the eslint rule unnecessarily, it looks less clean having the disabled rule. Worst case scenario is there is an actually an unforeseen error that is logged.

Reply:
@nedsalk Should we consider the scenario where the `child_process` starts successfully but terminates itself due to some internal error before the `cleanup` is called?

In this case, a try/catch could prevent a dirty exit.

Reply:
@arboleya I have added `'error'` and `'exit'` event handlers in https://github.com/FuelLabs/fuels-ts/pull/2718/commits/1babf7068d5c19f0b23eda05a664c72f5f49ddc9, and with it all termination modes are covered.

Reply:
It seems as though those changes are now causing the node to not be killed unfortunately 😓

Reply:
@maschad it seems to be something related to #2717. When I run the tests with `fuel-core 0.30.0` the nodes are killed, whereas with `fuel-core 0.31.0` they remain alive.

Reply:
@maschad  fixed in https://github.com/FuelLabs/fuels-ts/pull/2718/commits/823c33a83fd0474ead05e18889fc6578f8f7ec50.

Reply:
@nedsalk Have you tried what I said to see if the events will catch a termination happening from outside?

Reply:
You mentioned the `child_process` terminating itself, which is caught by the `'exit'` or `'error'` events. I have added a test case for external termination in https://github.com/FuelLabs/fuels-ts/pull/2718/commits/ebcf55c6aa0440a851ddabd6c828b4e64813bacc and it works.

Reply:
The test kills the child process from inside [of NodeJS env].

I suspect you'll get an error if it's killed from the outside [before the `cleanup` is called].

Reply:
Locally running the test below and running `kill -- -the_logged_pid` in a separate terminal passes the test, meaning the temporary directory got deleted.

```ts
  test(
    'external killing of node runs side-effect cleanup',
    async () => {
      const mkdirSyncSpy = vi.spyOn(fsMod, 'mkdirSync');

      const { pid } = await launchNode({ loggingEnabled: false });
      console.log(pid);
      expect(mkdirSyncSpy).toHaveBeenCalledTimes(1);
      const tempDirPath = mkdirSyncSpy.mock.calls[0][0];
      expect(tempDirPath).toBeTruthy();
      console.log(tempDirPath);
      // wait until cleanup finishes (done via events)
      await sleep(15000);
      expect(fsMod.existsSync(tempDirPath)).toBeFalsy();
    },
    { timeout: 100000 }
  );
```

Reply:
I didn't understand why you used a 16+ mins timeout on the test.

https://github.com/FuelLabs/fuels-ts/blob/ebcf55c6aa0440a851ddabd6c828b4e64813bacc/packages/account/src/test-utils/launchNode.test.ts#L218

---

I tried this repro on my end:

https://github.com/FuelLabs/fuels-ts/assets/26660/d9c4fd54-a992-4fa4-81db-9266d68dd801

Reply:
> I didn't understand why you used a 16+ mins timeout on the test.

Just an arbitrary number that doesn't interfere with the `sleep` call because the default is 5s. I wanted to give myself enough time to kill the node in a different terminal before the test finishes.

> I tried this repro on my end:

Okay, I understand now what you wanted. I updated the logic and added a test to cover that in https://github.com/FuelLabs/fuels-ts/pull/2718/commits/dc647cba6f1b97b2b5893aa351c289e73c4b6898.

Reply:
It's not about _me_, and given the task we're putting this utility through and planning for its future within our very own test suite (parallelism?, maybe even concurrency?), no amount of care and stability seems too much. You've even handled the specific error code. Well done, thanks for the attention to detail. 👏 🎯

Reply:
Out of curiosity, did you manage to trigger this case scenario from within the tests?

Or do the events always catch it from the perspective?

Reply:
> Well done, thanks for the attention to detail. 👏 🎯

Thanks! And thanks for the in-depth review!

> Out of curiosity, did you manage to trigger this case scenario from within the tests?

This scenario is covered by [this test](https://github.com/FuelLabs/fuels-ts/commit/e7abaeed48483675abe3c913397c8045a3e5e54a#diff-4a69805e7f6998db762f2618f4d54da8e4a4c6d69cbf8ba323b2e9b816096b64R216).

> Or do the events always catch the termination from this perspective?

Yes, the events are called whenever the child is terminated, be it from within the `process` or outside as you did in the video above.